### PR TITLE
Add Wasm AOT steps into WasmApp.targets

### PR DESCRIPTION
**Pull request by steveisok**
AOT can now be triggered through the libraries tests as well as the wasm samples.

Passing `/p:RunAOTCompilation=true` is all that is needed for the libraries test run. The property will be false (for now) by default.

### CODE REVIEW

File: src/mono/netcore/sample/wasm/console/Program.cs [hunk 1]
```diff
@@ -2,11 +2,12 @@
...
 
 using System;
+using System.Diagnostics.CodeAnalysis;
 using System.Threading.Tasks;
 
 public class Test
 {
-
+    [DynamicDependency(DynamicallyAccessedMemberTypes.All, "System.Runtime.InteropServices.JavaScript.Runtime", "System.Private.Runtime.InteropServices.JavaScript")] 
     public static async Task<int> Main(string[] args)
```
Comment (vargaz) [inline src/mono/netcore/sample/wasm/console/Program.cs:11]: This should probably be handled somewhere else.

File: src/mono/netcore/sample/wasm/console/WasmSample.csproj [hunk 2]
```diff
@@ -3,18 +3,16 @@
...
     <BuildDir>$(MSBuildThisFileDirectory)obj\$(Configuration)\wasm</BuildDir>
-    <AppDir>$(MSBuildThisFileDirectory)bin\$(Configuration)\AppBundle\</AppDir>
+    <AppBundleDir>$(MSBuildProjectDirectory)\bin\$(Configuration)\AppBundle\</AppBundleDir>
     <PublishTrimmed>true</PublishTrimmed>
     <TrimMode>link</TrimMode>
     <EnableTargetingPackDownload>false</EnableTargetingPackDownload>
-    <SelfContained>true</SelfContained>
     <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
-    <RunAnalyzers>false</RunAnalyzers>
     <RunAOTCompilation Condition="'$(RunAOTCompilation)' == ''">false</RunAOTCompilation>
```
Comment (vargaz) [inline src/mono/netcore/sample/wasm/console/WasmSample.csproj:16]: This is there to enable people to modify Program.cs during local development.

File: src/mono/netcore/sample/wasm/console/WasmSample.csproj [hunk 3]
```diff
@@ -37,79 +35,17 @@
...
-    <Exec Condition="'$(RunAOTCompilation)' == 'true'" Command="bash -c 'source $(EMSDK_PATH)/emsdk_env.sh &amp;&amp; emcc $(EmccFlags) --js-library $(MicrosoftNetCoreAppRuntimePackRidDir)native/src/library_mono.js --js-library $(MicrosoftNetCoreAppRuntimePackRidDir)native/src/binding_support.js --js-library $(MicrosoftNetCoreAppRuntimePackRidDir)native/src/dotnet_support.js --js-library $(MicrosoftNetCoreAppRuntimePackRidDir)native/src/pal_random.js driver.o pinvoke.o corebindings.o @(BundleAssemblies->'%(LlvmBitcodeFile)', ' ') libmono-ee-interp.a libmonosgen-2.0.a libmono-ilgen.a libmono-icall-table.a libSystem.Native.a libSystem.IO.Compression.Native.a libicuuc.a libicui18n.a -o dotnet.js'" IgnoreStandardErrorWarningFormat="true" WorkingDirectory="$(MSBuildThisFileDirectory)$(PublishDir)" />
-
+  <Target Name="PrepareForWasmBuild" BeforeTargets="WasmBuildApp">
     <PropertyGroup>
-      <WasmAppDir>$(AppDir)</WasmAppDir>
+      <WasmAppDir>$(AppBundleDir)</WasmAppDir>
+      <WasmAOTDir>$(MSBuildProjectDirectory)\$(PublishDir)\</WasmAOTDir>
       <WasmMainJSPath>$(MonoProjectRoot)\wasm\runtime-test.js</WasmMainJSPath>
```
Comment (vargaz) [inline src/mono/netcore/sample/wasm/console/WasmSample.csproj:42]: This should probably be named something like 'WasmObjDir' or 'WasmBuildDir'.

File: src/mono/wasm/build/WasmApp.targets [hunk 4]
```diff
@@ -1,13 +1,51 @@
...
+  <UsingTask TaskName="MonoAOTCompiler" AssemblyFile="$(MonoAOTCompilerTasksAssemblyPath)" />
   <UsingTask TaskName="WasmAppBuilder" AssemblyFile="$(WasmAppBuilderTasksAssemblyPath)" />
+  <UsingTask TaskName="WasmAOT" AssemblyFile="$(WasmAppBuilderTasksAssemblyPath)" />
   <UsingTask TaskName="WasmLoadAssembliesAndReferences" AssemblyFile="$(WasmAppBuilderTasksAssemblyPath)" />
 
   <Target Name="WasmBuildApp" AfterTargets="Publish">
     <Error Condition="'$(WasmMainAssemblyPath)' == ''" Text="%24(WasmMainAssemblyPath) property needs to be set" />
     <Error Condition="'$(WasmAppDir)' == ''" Text="%24(WasmAppDir) property needs to be set" />
+    <Error Condition="'$(RunAOTCompilation)' == 'true' and '$(WasmAOTDir)' == ''" Text="%24(WasmAOTDir) property needs to be set" />
     <Error Condition="!Exists('$(MicrosoftNetCoreAppRuntimePackRidDir)')" Text="MicrosoftNetCoreAppRuntimePackRidDir=$(MicrosoftNetCoreAppRuntimePackRidDir) doesn't exist" />
```
Comment (vargaz) [inline src/mono/wasm/build/WasmApp.targets:11]: This should be named something like WasmEnableAOT.

Comment (steveisok) [reply src/mono/wasm/build/WasmApp.targets:11]: `RunAOTCompilation` is a setting that applies to other platforms as well.

File: src/mono/netcore/sample/wasm/browser/WasmSample.csproj [hunk 5]
```diff
@@ -3,33 +3,53 @@
...
+    </ItemGroup>
+    <Message Text="Packaged ID: %(RuntimePack.PackageDirectory)" Importance="high" />
+  </Target>
+
+  <Target Name="PrepareForWasmBuild" BeforeTargets="WasmBuildApp">
     <PropertyGroup>
         <WasmAppDir>$(AppDir)</WasmAppDir>
-        <WasmMainAssemblyPath>bin\WasmSample.dll</WasmMainAssemblyPath>
+        <WasmAOTDir>$(MSBuildProjectDirectory)\$(PublishDir)\</WasmAOTDir>
+        <WasmMainAssemblyPath>$(MSBuildProjectDirectory)\bin\WasmSample.dll</WasmMainAssemblyPath>
```
Comment (mdh1418) [inline src/mono/netcore/sample/wasm/browser/WasmSample.csproj:42]: I think this can be removed, or the one on L44

**Review by vargaz — APPROVED** [review]

**Comment by vargaz** [comment]
We should merge this and fix it up later since it conflicts with other PRs.

File: src/tests/Common/wasm-test-runner/WasmTestRunner.proj [hunk 6]
```diff
@@ -10,6 +10,7 @@
     <NETCoreAppMaximumVersion>99.0</NETCoreAppMaximumVersion>
 
     <WasmAppBuilderTasksAssemblyPath>$(CORE_ROOT)\WasmAppBuilder\WasmAppBuilder.dll</WasmAppBuilderTasksAssemblyPath>
+    <MonoAOTCompilerTasksAssemblyPath>$(CORE_ROOT)\MonoAOTCompiler\MonoAOTCompiler.dll</MonoAOTCompilerTasksAssemblyPath>
```
Comment (naricc) [inline src/tests/Common/wasm-test-runner/WasmTestRunner.proj:13]: So, we haven't actually made the changes necessary for the runtime tests to use AOT, right? Is this property used by the libraries tests?

Comment (steveisok) [reply src/tests/Common/wasm-test-runner/WasmTestRunner.proj:13]: In theory, if you set the right inputs, the runtime tests could take advantage.  I didn't set that up... Only the libraries tests.

Comment (naricc) [reply src/tests/Common/wasm-test-runner/WasmTestRunner.proj:13]: I guess my question was; does it belong in this PR? Because right now nothing is using it.

Comment (steveisok) [reply src/tests/Common/wasm-test-runner/WasmTestRunner.proj:13]: I think we add enabling it for the runtime tests later.

File: src/mono/netcore/sample/wasm/browser/Makefile [hunk 7]
```diff
@@ -7,12 +7,16 @@ else
...
+DOTNET_RUN_AOT_COMPILATION_ARGS=/p:RunAOTCompilation=true
+endif
+
 CONFIG?=Release
 
 all: build
 
 build:
-	$(DOTNET) build $(DOTNET_Q_ARGS) /p:TargetArchitecture=wasm /p:TargetOS=Browser /p:Configuration=$(CONFIG) WasmSample.csproj
+	EMSDK_PATH=$(realpath $(TOP)/src/mono/wasm/emsdk) $(DOTNET) publish $(DOTNET_Q_ARGS) $(DOTNET_RUN_AOT_COMPILATION_ARGS) /p:Configuration=$(CONFIG) /p:TargetArchitecture=wasm /p:TargetOS=Browser $(MSBUILD_ARGS) WasmSample.csproj
```
Comment (lewing) [inline src/mono/netcore/sample/wasm/browser/Makefile:19]: Does this (realpath) work on OS/X without setting SHELL to a bash version newer than what it comes with?

File: src/mono/netcore/sample/wasm/browser/Program.cs [hunk 8]
```diff
@@ -2,12 +2,14 @@
...
 
 using System;
+using System.Diagnostics.CodeAnalysis;
 using System.Runtime.CompilerServices;
 
 namespace Sample
 {
     public class Test
     {
+        [DynamicDependency(DynamicallyAccessedMemberTypes.All, "System.Runtime.InteropServices.JavaScript.Runtime", "System.Private.Runtime.InteropServices.JavaScript")]
```
Comment (marek-safar) [inline src/mono/netcore/sample/wasm/browser/Program.cs:12]: Yep, this is wrong place.

File: Directory.Build.targets [hunk 9]
```diff
@@ -21,4 +21,4 @@
     -->
     <NETCoreAppMaximumVersion>$(MajorVersion).$(MinorVersion)</NETCoreAppMaximumVersion>
   </PropertyGroup>
-</Project>
+</Project>
```
Comment (akoeplinger) [inline Directory.Build.targets:24]: unrelated change.

File: src/mono/netcore/sample/wasm/browser/Makefile (hunk 7 above)
Comment (akoeplinger) [reply src/mono/netcore/sample/wasm/browser/Makefile:19]: we should only set EMSDK_PATH if it's not yet set (so I can use my own emsdk). If you do that as a Makefile variable you can easily resolve it to a full path with the `abspath` Makefile function:

```
EMSDK_PATH?=$(abspath $(TOP)/src/mono/wasm/emsdk)
```

File: src/mono/netcore/sample/wasm/console/Makefile [hunk 10]
```diff
@@ -17,7 +17,7 @@ CONFIG?=Release
 all: build
 
 build:
-	EMSDK_PATH=$(realpath $(TOP)/src/mono/wasm/emsdk) $(DOTNET) publish $(DOTNET_Q_ARGS) $(DOTNET_RUN_AOT_COMPILATION_ARGS) /p:TargetArchitecture=wasm /p:TargetOS=Browser /p:Configuration=$(CONFIG) WasmSample.csproj
+	EMSDK_PATH=$(realpath $(TOP)/src/mono/wasm/emsdk) $(DOTNET) publish $(DOTNET_Q_ARGS) $(DOTNET_RUN_AOT_COMPILATION_ARGS) /p:TargetArchitecture=wasm /p:TargetOS=Browser /p:Configuration=$(CONFIG) $(MSBUILD_ARGS) WasmSample.csproj
```
Comment (akoeplinger) [inline src/mono/netcore/sample/wasm/console/Makefile:20]: same comment about EMSDK_PATH here.

File: src/mono/wasm/build/WasmApp.targets [hunk 11]
```diff
@@ -1,31 +1,84 @@
...
+      <_WasmAssemblies Remove="@(_WasmAssemblies)" />
+      <_WasmAssemblies Include="@(_TmpWasmAssemblies)" />
     </ItemGroup>
+  </Target>
+
+  <Target Name="_BeforeWasmBuildApp">
+      <Error Condition="'$(WasmMainAssemblyPath)' == ''" Text="%24(WasmMainAssemblyPath) property needs to be set" />
+      <Error Condition="'$(WasmAppDir)' == ''" Text="%24(WasmAppDir) property needs to be set" />
+      <Error Condition="'$(RunAOTCompilation)' == 'true' and '$(WasmAOTDir)' == ''" Text="%24(WasmAOTDir) property needs to be set" />
+      <Error Condition="!Exists('$(MicrosoftNetCoreAppRuntimePackRidDir)')" Text="MicrosoftNetCoreAppRuntimePackRidDir=$(MicrosoftNetCoreAppRuntimePackRidDir) doesn't exist" />
```
Comment (akoeplinger) [inline src/mono/wasm/build/WasmApp.targets:67]: Hm I wonder if we still need to set this explicitly in the projects. We could resolve all the other paths that currently use it to use `%(RuntimePack.PackageDirectory)` instead. That's also basically what we'll need to do for "real" usage of this target outside of dotnet/runtime.

File: src/mono/wasm/build/WasmApp.targets [hunk 12]
```diff
@@ -38,9 +91,43 @@
     <CallTarget Targets="_GenerateRunV8Script" Condition="'$(WasmGenerateRunV8Script)' == 'true'" />
   </Target>
 
+  <Target Name="_AfterWasmBuildApp" AfterTargets="_WasmBuildApp">
+    <!-- Run mono-cil-strip on the assemblies -->
+    <!-- TODO: replace this with a linker task -->
```
Comment (akoeplinger) [inline src/mono/wasm/build/WasmApp.targets:96]: do you remember if we created an issue for that?

File: src/mono/netcore/sample/wasm/browser/WasmSample.csproj (hunk 5 above)
Comment (akoeplinger) [reply src/mono/netcore/sample/wasm/browser/WasmSample.csproj:43]: Yeah this looks fishy.

**Comment by steveisok** [comment]
The remaining feedback will be addressed in a follow up PR
